docs/architecture.md

```diff
@@ -0,0 +1,473 @@
+
+```rust
+trait PeerToPeer {
+    type SealedHeaderResponse: NetworkData<SealedFuelBlockHeader>;
+    type BlockResponse: NetworkData<Vec<SealedFuelBlock>>;
+    type GossipedBlockHeader: NetworkData<SealedFuelBlockHeader>;
+    
+    async fn fetch_best_network_block_header() -> Result<Self::SealedHeaderResponse>;
+    async fn fetch_blocks(query: Range<BlockHeight>) -> Result<Self::BlockResponse>;
+    // punish the sender for providing an invalid block header
+    fn report_invalid_block_header(invalid_header: &Self::SealedHeaderResponse) -> Result<()>;
+    // punish the sender for providing a set of blocks that aren't valid
+    fn report_invalid_blocks(invalid_blocks: &Self::BlockResponse) -> Result<()>;
+    // await a newly produced block from the network (similar to stream.next())
+    async fn next_gossiped_block_header() -> Result<Self::SealedFuelBlockHeader>;
+    // notify the p2p network whether to continue gossiping this message to others or
+    // punish the peer that sent it
+    fn notify_gossip_block_validity(message: &Self::GossipedBlockHeader, validity: GossipValidity);
+}
+
+// Generic wrapper for data received from the network
+pub trait NetworkData<D> {
+    // Transfer ownership of the provided data without consuming other message metadata
+    fn take_data(&mut self) -> Option<D>;
+}
+
+pub enum GossipValidity {
+    // Report whether the gossiped message is valid and safe to rebroadcast
+    Accept,
+    // Ignore the received message and prevent further gossiping
+    Ignore,
+    // Punish the gossip sender for providing invalid 
+    // (or malicious) data and prevent further gossiping
+    Invalid,
+}
+
+trait BlockImporter {
+    // commit a sealed block to the chain
+    async fn commit(block: SealedFuelBlock) -> Result<()>;
+}
+
+trait Consensus {
+    // check with the consensus layer whether a block header passes consensus rules
+    async fn validate_sealed_block_header(block: SealedFuelBlockHeader) -> Result<()>;
+}
+
+trait Database {
+    // get current fuel blockchain height
+    fn get_current_height() -> Result<BlockHeight>;
+}
+```
+#### Adapters: fuel_core::service::adapters::fuel_sync
+_N/A_
+
+### P2P
+
+#### Ports: fuel_p2p::ports
+```rust
+trait Database {
+    // used when other peers are syncing
+    fn get_sealed_block(height: BlockHeight) -> Result<SealedFuelBlock>; 
+}
+```
+
+#### Adapters: fuel_core::service::adapters::peer_to_peer
+```rust
+impl fuel_poa_consensus::ports::PeerToPeer for Service<P2pService> {
+    // insert impl here
+}
+
+impl fuel_sync::ports::PeerToPeer for Service<P2pService> {
+    // insert impl here
+}
+
+impl fuel_txpool::ports::PeerToPeer for Service<P2pService> {
+    // insert impl here
+}
+
+```
+
+### Transaction Pool
+
+#### Ports: fuel_txpool::ports
+```rust
+trait PeerToPeer {
+    type GossipedTransaction: NetworkData<Transaction>;
+    // Gossip broadcast a transaction inserted via API.
+    async fn broadcast_transaction(transaction: Transaction);
```

Comment:
What does this async actually wait for? There is no error here so it can't really wait for the broadcast to send or be received (which is failable). Also similar concerns to my above question on broadcast.

Reply:
true, this probably doesn't need to be async

Reply:
Maybe only because of overloading the broadcast task, see my other comment https://github.com/FuelLabs/fuel-core/pull/663#discussion_r995207775

Reply:
libp2p actually uses sync interface to put a broadcast request onto the queue. There's likely some parameters within the library we have to tune to address the issues you raised, but that doesn't need to percolate up through these interfaces luckily.

Reply:
It would probably be good to have a comment to say what it means when this function returns though as it affects how it should be used.

Reply:
hmm looking closer, it actually looks libp2p directly does the full broadcast operation (along with all the encoding and networking) from this sync call. Our p2p service has a separately spawned task that manages these broadcast requests, so that will likely require async at this level.

Reply:
Yeh probably to prevent spawning too many tasks